### PR TITLE
Set the dimensions of the various layers at their creation

### PR DESCRIPTION
**Pull request by calixteman**
- Use a unique helper function in display/display_utils.js;
- Move those dimensions in the css' side.

### CODE REVIEW

File: src/display/display_utils.js [hunk 1]
```diff
@@ -614,6 +614,45 @@ function getCurrentTransformInverse(ctx) {
...
+  const widthStr = `calc(var(--scale-factor) * ${pageWidth}px)`;
+  const heightStr = `calc(var(--scale-factor) * ${pageHeight}px)`;
+
+  if (!mustFlip || viewport.rotation % 180 === 0) {
+    style.width = widthStr;
+    style.height = heightStr;
+  } else {
+    style.width = heightStr;
+    style.height = widthStr;
+  }
```
Comment (Snuffleupagus) [inline src/display/display_utils.js:649]: Can we perhaps avoid duplicating the rotation-handling with something like the below, which should be a tiny bit less code overall?

```suggestion
  if (viewport.viewBox) {
    const [pageLLx, pageLLy, pageURx, pageURy] = viewport.viewBox;
    const pageWidth = pageURx - pageLLx;
    const pageHeight = pageURy - pageLLy;
    const { style } = div;
    const widthStr = `calc(var(--scale-factor) * ${pageWidth}px)`;
    const heightStr = `calc(var(--scale-factor) * ${pageHeight}px)`;

    if (!mustFlip || viewport.rotation % 180 === 0) {
      style.width = widthStr;
      style.height = heightStr;
    } else {
      style.width = heightStr;
      style.height = widthStr;
    }
  }
```

**Comment by Snuffleupagus** [comment]
I'm holding off on testing/review this a little bit, since it seems like a good idea to not land this and the textLayer changes in the same mozilla-central update to make finding/handling any potential regressions easier.

**Comment by Snuffleupagus** [comment]
*Let's check that all tests pass before I attempt to test/review this in more detail.*

/botio test

**Comment by Snuffleupagus** [comment]
One observation here is that in some PDF documents the page-containers, and all of their layers, are no longer guaranteed to have *integer* dimensions. Hopefully that's not really an issue anywhere in the viewer, e.g. in the scrolling and visibility-checking code, but it's a sort-of "noticeable change" and one that I cannot immediately tell if it'll always be safe (given the age of this viewer code).
*Ninja-edit:* Or is the *browser* perhaps rounding fractional dimensions internally already, making this entire point moot?

**Comment by calixteman** [comment]
I'd say that the problem of non-integer dimensions exists for a while because of the `%` unit which like lead to non-integer numbers at the end, so I'm pretty confident.
The only issues I could imagine about rounding errors would be to have some containers with a small black line between them (exactly like the issue fixed by https://github.com/mozilla/pdf.js/pull/14853), but in the pdf.js context we don't have such adjacent containers.
@emilio, wdyt ?

**Comment by Snuffleupagus** [comment]
> I'd say that the problem of non-integer dimensions exists for a while because of the `%` unit which like lead to non-integer numbers at the end, so I'm pretty confident.

The difference is that previously it never happened for the `page`-elements, which means that there's not way it could've affected e.g. scrolling and visibility calculations.
Now you're changing that, hence my worry about this possibly breaking some (old) edge-cases. Note that while methods such as e.g. `Element.clientWidth` always return integer dimensions, however using the Inspector (in the dev-tools) fractional dimensions are displayed (for some PDF documents).

**Comment by calixteman** [comment]
Ah yes that's true.
I've a wip locally to use an `IntersectionObserver`  for the visibility stuff.
In looking at the code in `getVisibleElements`, I don't really see what could be wrong (the various clientXXX values are rounded compared to what we have in `getComputedStyle(...).width`).
Anyway, let's try and we'll see/fix the problems if any (I know I'm usually too much confident).
But as usual, I really rely on your judgement: you're the old wise man here so don't hesitate to try to figure out something which could break the rendering.

**Comment by Snuffleupagus** [comment]
> I've a wip locally to use an `IntersectionObserver` for the visibility stuff.

As far as I understand things, that one is *async*. Please keep in mind that there's a bunch of situations where we need to access the visible elements *synchronously*, so I'm not sure how well that'd work in general.
Furthermore given the all of the viewer-related changes recently it seems that making such large, and potentially invasive, re-factoring of old and somewhat poorly tested code carries a way too large risk here. (Since there's just so many different cases to consider...)
Hence, I'm sorry to say that such a change mostly just scares me here :-(

> In looking at the code in `getVisibleElements`, I don't really see what could be wrong (the various clientXXX values are rounded compared to what we have in `getComputedStyle(...).width`).

Could it not potentially lead to *the wrong page* being considered as the most visible, thus affecting things such as rendering prioritization in some edge-cases, exactly because of the rounding that you mention!?
I suppose that we can't easily e.g. `floor` the values computed by the new `setLayerDimensions` function, since that ought to prevent any issues?

**Comment by calixteman** [comment]
> > I've a wip locally to use an `IntersectionObserver` for the visibility stuff.
> 
> As far as I understand things, that one is _async_. Please keep in mind that there's a bunch of situations where we need to access the visible elements _synchronously_, so I'm not sure how well that'd work in general. Furthermore given the all of the viewer-related changes recently it seems that making such large, and potentially invasive, re-factoring of old and somewhat poorly tested code carries a way too large risk here. (Since there's just so many different cases to consider...) Hence, I'm sorry to say that such a change mostly just scares me here :-(
> 

Yep I know it's risky and I'll try to figure out a way to reduce the risk as much as possible but this patch is not a priority for the moment.

> > In looking at the code in `getVisibleElements`, I don't really see what could be wrong (the various clientXXX values are rounded compared to what we have in `getComputedStyle(...).width`).
> 
> Could it not potentially lead to _the wrong page_ being considered as the most visible, thus affecting things such as rendering prioritization in some edge-cases, exactly because of the rounding that you mention!? I suppose that we can't easily e.g. `floor` the values computed by the new `setLayerDimensions` function, since that ought to prevent any issues?

In the future, we can use:
https://developer.mozilla.org/en-US/docs/Web/CSS/round

I think `Math.floor(x) === Math.round(x - 0.5)` so since it appears that `clientWidth === round(getComputedStyle(..).width)` a way to floor would be to just subtract 0.5 in the `calc`.

**Comment by Snuffleupagus** [comment]
> In the future, we can use:
> https://developer.mozilla.org/en-US/docs/Web/CSS/round

What's the status of that feature, w.r.t. being enabled by default?
Could we perhaps enable that *unconditionally* for the Firefox PDF Viewer specifically, since I seem to recall that there's existing cases where we "force-enable" features in that case?

**Comment by calixteman** [comment]
> > In the future, we can use:
> > https://developer.mozilla.org/en-US/docs/Web/CSS/round
> 
> What's the status of that feature, w.r.t. being enabled by default? Could we perhaps enable that _unconditionally_ for the Firefox PDF Viewer specifically, since I seem to recall that there's existing cases where we "force-enable" features in that case?

For now it's only in nightly:
https://searchfox.org/mozilla-central/rev/a44deb21744de6269329b05461c55488a147f95e/modules/libpref/init/StaticPrefList.yaml#8028
Considering that the pref is used in the css parser:
https://searchfox.org/mozilla-central/rev/a44deb21744de6269329b05461c55488a147f95e/servo/components/style/values/specified/calc.rs#789
I'm not sure it's possible to force the pref when the principal is the pdf.js one, I'll ask.

**Comment by Snuffleupagus** [comment]
> And to clarify my goal isn't to change just because changing is kind of good/cool, but I just want to simplify things because I want to not be scared by any changes and I want to improve performance in general.

Of course I understand that, and it's obviously a good goal!

(With *most* other long-time core PDF.js developers no longer active in the project, I'm really struggling to try and keep up with all changes here. Perhaps I also tend to be needlessly worried/pessimistic about some larger changes, really sorry about that!)

File: src/display/display_utils.js [hunk 2]
```diff
@@ -614,6 +614,52 @@ function getCurrentTransformInverse(ctx) {
...
+    const pageWidth = pageURx - pageLLx;
+    const pageHeight = pageURy - pageLLy;
+    const { style } = div;
+
+    // The final `- 0.5px` are here to help to finally have some clientWidth and
+    // clientHeight which are the floored version of:
+    //  - var(--scale-factor) * ${pageWidth}px
+    //  - var(--scale-factor) * ${pageHeight}px
+    const widthStr = `calc(var(--scale-factor) * ${pageWidth}px - 0.5px)`;
+    const heightStr = `calc(var(--scale-factor) * ${pageHeight}px - 0.5px)`;
```
[the remark this answers is not in the document]
Comment (Snuffleupagus) [reply src/display/display_utils.js:641]: Another option would be that we just leave the fractional dimensions as-is, as in the original patch, and then land this and see how it goes!

**Comment by calixteman** [comment]
Yep, @emilio told me that we could try to enable it by default in 110.
Anyway the `-0.5px` does the trick so why not using it if you prefer.
Another thing I thought about: you told that `IntersectionObserve` is async and that's correct but right now it's the same
Each scroll event just add either a callback in `window.requestAnimationFrame` and with my setup it seems that it's called every 16ms (likely 60 fps) or just do nothing if a callback is already there. 
When the callback is called then `PDFViewer::update` is called and then `getVisibleElements`. Between the scrollEvent which triggered the update and the call to `getVisibleElements` some things could have happened.
So if I understand correctly, some scroll events are just skipped which means that likely few pixels are in the area when we try to compute the visibility.
And even the `window.requestAnimationFrame` is not exactly called every 16ms: it depends on the event queue...

So my feeling is that this patch will bring a potential new inaccuracy to something which is already inaccurate but it doesn't mean that it'll be more inaccurate (errors can compensate themselves sometimes).
I just tried to dump the `hiddenHeight` of the first visible element and the values dumped when scrolling are never the same from a run to an other.
Anyway, we can ask to QA to play with their mouse wheel in different configuration to try to catch something wrong.

**Comment by calixteman** [comment]
> > Another thing I thought about: you told that `IntersectionObserve` is async and that's correct but right now it's the same
> > Each scroll event just add either a callback in `window.requestAnimationFrame` and with my setup it seems that it's called every 16ms (likely 60 fps) or just do nothing if a callback is already there.
> > When the callback is called then `PDFViewer::update` is called and then `getVisibleElements`.
> 
> Please note that my point was/is _not_ actually about that particular code-path, but rather about the _various cases_ throughout the `PDFViewer`-implementation where we directly call `this._getVisiblePages()` in order to _synchronously_ find out which pages are currently visible.

Ok, so what could happen ?
Some pages in whatever view could be out of the viewport because of rounding issues, so would it mean that right now we're seeing only 1 pixel of them ? because finally floor vs round is really a matter of 1px. 
Will we be wrong in not rendering this newly invisible page ? The only difference I see finally is that when the user will scroll then the page would have to be rendered when right now it's already rendered. But that means that the viewport and the page size (+its border) have the exact dimensions required to exactly have 1px of page 2 in the viewport which is unlikely.

Again, why not using the -0.5px trick in order to have exactly the same behaviour as now ?

**Comment by calixteman** [comment]
And if there are some bugs, we'll fix them as usual :)

**Comment by Snuffleupagus** [comment]
Sorry if it was unclear, https://github.com/mozilla/pdf.js/pull/15770#issuecomment-1343026984 was only about your previous suggestion to maybe use `IntersectionObserve` in the viewer (not about this patch as such).

**Comment by calixteman** [comment]
> Sorry if it was unclear, [#15770 (comment)](https://github.com/mozilla/pdf.js/pull/15770#issuecomment-1343026984) was only about your previous suggestion to maybe use `IntersectionObserve` in the viewer (not about this patch as such).

Ok I thought it was about this patch.
I just used the `IntersectionObserver` thing as an example to show why it's right now inaccurate.
And about the `IntersectionObserver` thing, we'll see that later (next year), it's absolutely not a priority.

File: src/display/editor/annotation_editor_layer.js [hunk 3]
```diff
@@ -16,6 +16,8 @@
 /** @typedef {import("./editor.js").AnnotationEditor} AnnotationEditor */
 // eslint-disable-next-line max-len
 /** @typedef {import("./tools.js").AnnotationEditorUIManager} AnnotationEditorUIManager */
+
```
Comment (Snuffleupagus) [inline src/display/editor/annotation_editor_layer.js:19]: Nit: Unnecessary new-line.

**Review by Snuffleupagus — APPROVED** [review]
OK let's try this, since I'm assuming that this won't land in mozilla-central until the 110-cycle thus giving us some time to find problems.
But please file a follow-up issue about using CSS `round` as soon as that one becomes generally available in Firefox!

**Comment by calixteman** [comment]
> OK let's try this, since I'm assuming that this won't land in mozilla-central until the 110-cycle thus giving us some time to find problems. But please file a follow-up issue about using CSS `round` as soon as that one becomes generally available in Firefox!

Sorry to ask again, but why not use the `-0.5px` trick ?

**Comment by Snuffleupagus** [comment]
> Sorry to ask again, but why not use the `-0.5px` trick ?

I don't fully understand how that helps, since it still leads to non-integer dimensions for these layers. Hence we have a situation where the actual dimensions differ from what e.g. `clientWidth`/`clientHeight` returns, as far as I can tell.

**Comment by calixteman** [comment]
> > Sorry to ask again, but why not use the `-0.5px` trick ?
> 
> I don't fully understand how that helps, since it still leads to non-integer dimensions for these layers. Hence we have a situation where the actual dimensions differ from what e.g. `clientWidth`/`clientHeight` returns, as far as I can tell.

It'd help to at least have the same values as now for those dimensions.
Do we use somewhere the "real" (e.g. `getComputedStyle(div).width` or something similar) dimensions of the page div ?
I just noticed that the values `pageView.width`/`pageView.height` are potentially non-integers (it depends of the scale factor), so we already have a potential mismatch between `clientX` and `pageView.X`.

**Pull request merged**